packages/react/src/jsx.tsx

```diff
@@ -13,27 +12,20 @@ import {
   StateDataState,
 } from './template'
 
-// tbd: refactor and unify this
-type AllowedDataStates =
-  | ServerEventDataState
-  | ActionDataState
-  | ReferencedDataState
-  | StateDataState
-  | ((args: any) => any)
+export type UI = ReactElement<Props> | ServerComponent
 
-type Props = Record<string, JSONValue | AllowedDataStates> & {
-  children: DataState
+type ServerComponent = ComponentDataState<ServerEventDataState>
+type Props = ServerComponent['props'] & {
```

Comment:
refactor: do not duplicate "prop" types that can be defined for components on the server. Instead, reference props from the real type.